### PR TITLE
titus_task_relocation_api IDL update

### PR DESCRIPTION
**Pull request by tbak**
### Description of the Change

A major rewrite of the original version of the titus_task_relocation_api IDL.

### CODE REVIEW

File: src/main/proto/netflix/titus/titus_job_api.proto [hunk 1]
```diff
@@ -187,16 +187,28 @@ message JobDisruptionBudget {
...
         AvailableTasksLimit availableTasksLimit = 2;
-        EvictionLimit relocationLimit = 3;
+        RelocationLimit relocationLimit = 3;
     }
+
+    /// Defines the relocation delay. A user is notified about the
+    //  relocation plan, and may choose to handle the relocation itself.
+    //  If the contain is still running after the delay passes, it is
+    //  moved by the task relocation process.
+    uint64 relocationDelayMs = 4;
```
Comment (corindwyer) [inline src/main/proto/netflix/titus/titus_job_api.proto]: Remove this and just add a `Self Managed` policy or mechanism?

File: src/main/proto/netflix/titus/titus_job_api.proto [hunk 2]
```diff
@@ -187,16 +187,28 @@ message JobDisruptionBudget {
...
     }
 
-    /// Maximum number of times a task can be evicted.
-    message EvictionLimit {
+    /// Maximum number of times a task can be relocated.
+    message RelocationLimit {
         uint32 limit = 1;
     }
 
     oneof Policy {
```
Comment (corindwyer) [inline src/main/proto/netflix/titus/titus_job_api.proto]: Convert the concept to a rate or something that allows a number of terminations per time period?

File: src/main/proto/netflix/titus/titus_task_relocation_api.proto [hunk 3]
```diff
@@ -17,62 +17,101 @@ option go_package = "titus";
...
-/// A single relocation plan.
-message RelocationPlan {
-    /// (Required) A list of task ids to terminate.
-    repeated string candidateTaskIds = 1;
+message TaskRelocationPlan {
+    /// (Required) Id of a task to relocate.
+    string taskId = 1;
+
+    /// (Required) Relocation reason code. Currently defined reason codes are:
+    //  * 'migrationFromDeprecatedAgent'
```
Comment (corindwyer) [inline src/main/proto/netflix/titus/titus_task_relocation_api.proto]: Remove reason codes until we figure out the different reasons? These names are based on older concepts

File: src/main/proto/netflix/titus/titus_task_relocation_api.proto [hunk 4]
```diff
@@ -17,62 +17,101 @@ option go_package = "titus";
...
+    //  * 'evacuationFromQuarantinedAgent'
+    //  * 'defragmentation'
+    string reasonCode = 2;
+
+    /// (Optional) Human readable text providing more details on why the relocation is needed.
+    string reasonMessage = 3;
+
+    /// (Optional) The earliest time at which the relocation may happen. Until that time, an owner of the task
+    //   may chose to do the relocation themselve.
+    uint64 relocationAfter = 4;
```
Comment (corindwyer) [inline src/main/proto/netflix/titus/titus_task_relocation_api.proto]: This name is confusing.

File: src/main/proto/netflix/titus/titus_task_relocation_api.proto [hunk 5]
```diff
@@ -17,62 +17,101 @@ option go_package = "titus";
...
+    /// (Required) Execution result. Each task relocation state is associated with a different set of the reason codes.
+    string reasonCode = 2;
+
 }
 
 /// An execution status for a single relocation plan.
-message ExecutionStatus {
-    /// Task ids from a relocation plan, that were successfully evicted.
-    repeated string terminatedTaskIds = 1;
+message TaskRelocationResult {
```
Comment (corindwyer) [inline src/main/proto/netflix/titus/titus_task_relocation_api.proto]: TaskRelocationExecution?

File: src/main/proto/netflix/titus/titus_task_relocation_api.proto [hunk 6]
```diff
@@ -17,62 +17,101 @@ option go_package = "titus";
...
+    /// (Required) Executed task relocation plan.
+    TaskRelocationPlan taskRelocationPlan = 1;
 
-    // Map with key values containing the task ids from a relocation plan, that could not be evicted due to disruption
-    // budget or other constraints. Map value is the failure reason.
-    map<string, string> taskTerminationFailures = 2;
+    /// (Required) An orderd list of relocation attempts, starting from the oldest one to the latest.
```
Comment (corindwyer) [inline src/main/proto/netflix/titus/titus_task_relocation_api.proto]: orderd -> ordered

File: src/main/proto/netflix/titus/titus_task_relocation_api.proto [hunk 7]
```diff
@@ -17,62 +17,101 @@ option go_package = "titus";
...
-    RelocationPlan plan = 1;
-    ExecutionStatus status = 2;
+message TaskRelocationPlans {
+    repeated TaskRelocationPlan plans = 1;
 }
 
-message RelocationHistoryRequest {
-    /// (Required) Number of the previously executed relocation plans to return.
-    uint32 count = 1;
+message TaskRelocationResults {
```
Comment (corindwyer) [inline src/main/proto/netflix/titus/titus_task_relocation_api.proto]: TaskRelocationResults -> TaskRelocationExecutions

File: src/main/proto/netflix/titus/titus_task_relocation_api.proto [hunk 8]
```diff
@@ -17,62 +17,101 @@ option go_package = "titus";
...
+        TaskRelocationPlanEvent taskRelocationPlanEvent = 1;
+        TaskRelocationResultEvent taskRelocationResultEvent = 2;
     }
 }
 
 service TaskRelocationService {
 
     /// Get the current relocation plan.
-    rpc GetCurrent (google.protobuf.Empty) returns (RelocationResult) {
+    rpc GetCurrentTaskRelocationPlans (google.protobuf.Empty) returns (TaskRelocationPlans) {
```
Comment (corindwyer) [inline src/main/proto/netflix/titus/titus_task_relocation_api.proto]: Add query to filter for specific plans

File: src/main/proto/netflix/titus/titus_job_api.proto (hunk 1 above)
Comment (tbak) [reply src/main/proto/netflix/titus/titus_job_api.proto]: Done

File: src/main/proto/netflix/titus/titus_job_api.proto (hunk 2 above)
Comment (tbak) [reply src/main/proto/netflix/titus/titus_job_api.proto]: I have introduced a new oneof type `Rate`.

File: src/main/proto/netflix/titus/titus_task_relocation_api.proto (hunk 3 above)
Comment (tbak) [reply src/main/proto/netflix/titus/titus_task_relocation_api.proto]: Ok, removed

File: src/main/proto/netflix/titus/titus_task_relocation_api.proto (hunk 4 above)
Comment (tbak) [reply src/main/proto/netflix/titus/titus_task_relocation_api.proto]: `relocationTime`?

File: src/main/proto/netflix/titus/titus_task_relocation_api.proto (hunk 5 above)
Comment (tbak) [reply src/main/proto/netflix/titus/titus_task_relocation_api.proto]: Changed

File: src/main/proto/netflix/titus/titus_task_relocation_api.proto (hunk 6 above)
Comment (tbak) [reply src/main/proto/netflix/titus/titus_task_relocation_api.proto]: Fixed

File: src/main/proto/netflix/titus/titus_task_relocation_api.proto (hunk 7 above)
Comment (tbak) [reply src/main/proto/netflix/titus/titus_task_relocation_api.proto]: Fixed

**Pull request merged**